### PR TITLE
[Proposal] Add a helper class for Node.traverse()

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Refactoring

### Purpose
- In sphinx, there are many small inline functions to pick up nodes conditionally.
- I'd like to add a helper class named `NodeMatcher`.
- It can search nodes simply.

### CODE REVIEW

File: sphinx/util/nodes.py [hunk 1]
```diff
@@ -33,6 +34,49 @@
...
+        # type: (nodes.Node, Any) -> None
+        self.classes = classes
+        self.attrs = attrs
+
+    def match(self, node):
+        # type: (nodes.Node) -> bool
+        try:
+            if self.classes and not isinstance(node, self.classes):
+                return False
+            else:
```
Comment (cocoatomo) [inline sphinx/util/nodes.py]: This if-else statement might be in early-return form as following:

```
if self.classes and not isinstance(node, self.classes):
    return False

for key, value in self.attrs.items():
    if key not in node:
...
```

Early-return form reduce depth of indentation and make it easier to read source codes.

File: sphinx/util/nodes.py [hunk 2]
```diff
@@ -33,6 +34,49 @@
 caption_ref_re = explicit_title_re  # b/w compat alias
 
 
+class NodeMatcher(object):
+    """A helper class for Node.traverse().
```
Comment (cocoatomo) [inline sphinx/util/nodes.py:38]: Descriptions about `Any` value of some attribute are necessary.

**Review by cocoatomo — CHANGES_REQUESTED** [review]
The topic of early-return is about code style, so it is an approval.
But the topic of documentation is mandatory.

File: sphinx/util/nodes.py (hunk 1 above)
Comment (tk0miya) [reply sphinx/util/nodes.py]: Thanks, I think so too.

File: sphinx/util/nodes.py (hunk 2 above)
Comment (tk0miya) [reply sphinx/util/nodes.py:38]: Absolutely!

**Review by shimizukawa — APPROVED** [review]
LGTM!!

**Comment by tk0miya** [comment]
Merged. Thank you for reviewing.

**Pull request merged**